### ISSUE TITLE
attribute test now returning 2 elements

### ISSUE THREAD

**Issue by dagda1** [airbnb/enzyme#1406]
I have just upgraded from enzyme 2 to 3 and the following test that previously returned 1 element now returns 2:
```
  it('should find element', () => {
    const wrapper = mount(<MyComponent/>);

    expect(wrapper.find('[name="element"]')).toHaveLength(1);
  });
```

I found [this](https://github.com/airbnb/enzyme/blob/master/docs/guides/migration-from-2-to-3.md#element-referential-identity-is-no-longer-preserved) link that seems to suggest that this is the expected behaviour.

It does not say why this is good or how should I now be writing my tests?

**Comment by ljharb** [comment]
This is good for overall consistency throughout the API; if you do `wrapper.hostNodes().find` you'll get the same behavior as in enzyme 2.

**Comment by dagda1** [comment]
Thank you for the response:

`wrapper.hostNodes().find('[name="element"]'))` returns 0 in my case.

Could this be a potential bug?

**Comment by dagda1** [comment]
So there definitely something weird with attribute selectors.

I have a component that renders a select, if I I do:

`wrapper.find('select').length` then it returns 1.

I was testing the aria tags on this component and:

`expect(wrapper.find('[aria-invalid=true]')).toHaveLength(1);`

Fails because 2 is being returned.

**Comment by ljharb** [comment]
Can you share `wrapper.debug()` and ``wrapper.find(‘[aria-invalid=true]’).debug()`?

**Comment by dagda1** [comment]
`wrapper.debug()` returns:
```
      <FormControlWrapper options={{...}} label="label" invalid={true} strong={true}>
        <div className="form-group">
          <Label id="ctrl5-label" htmlFor="ctrl5" required={[undefined]} strong={true} doubleLabelMargin={true} dataSelector={[undefined]}>
            <label htmlFor="ctrl5" id="ctrl5-label" className="undefined" data-selector={[undefined]}>
              label
            </label>
          </Label>
          <div className="input-wrapper    ">
            <Component id="ctrl5" name={[undefined]} invalid={true} aria-invalid={true} required={[undefined]} aria-describedby="ctrl5-error" options={{...}} defaultLabel="" filterOptions={[Function]}>
              <div className={[undefined]}>
                <select value={[undefined]} className="undefined" id="ctrl5" name={[undefined]} aria-invalid={true} required={[undefined]} aria-describedby="ctrl5-error">
                  <option value="" />
                  <option>
                    1
                  </option>
                  <option>
                    2
                  </option>
                  <option>
                    3
                  </option>
                </select>
              </div>
            </Component>
          </div>
          <div id="ctrl5-error" className="form-group" aria-hidden={false} role="alert" />
        </div>
      </FormControlWrapper>
```
`wrapper.find('[aria-invalid=true]')` returns:
`ReactWrapper { length: 2 }`

**Comment by dagda1** [comment]
The component is an hoc:

```
import * as React from 'react';
import { Label } from '../../atoms/Label/Label';
import { Error } from '../../atoms/ErrorLabel/ErrorLabel';
import { prefixId, gridItemAdapter, BreakPointProps } from '../../../util';
import * as cs from 'classnames';

const styles = require('./FormControl.scss');

export type Component<T> =
  | React.StatelessComponent<T>
  | React.ComponentClass<T>;

export interface FormControlWrapperProps extends BreakPointProps {
  additionalLabel?: React.ReactNode;
  className?: string;
  doubleLabelMargin?: boolean;
  errorDataSelector?: string;
  errorMessage?: string;
  highlight?: boolean;
  id?: string;
  invalid?: boolean;
  label: string;
  name?: string;
  required?: boolean;
  strong?: boolean;
  width?: any;
}

export function FormControl<T>(
  Comp: Component<T>
): React.ComponentClass<FormControlWrapperProps & T> {
  return class FormControlWrapper extends React.Component<
    FormControlWrapperProps & T
  > {
    id: string;

    static defaultProps = {
      strong: true
    } as any;

    constructor(props: FormControlWrapperProps & T) {
      super(props);

      this.id = this.props.id || this.props.name || prefixId();
    }

    render() {
      const {
        invalid,
        name,
        label,
        strong,
        doubleLabelMargin,
        errorDataSelector,
        errorMessage,
        className,
        additionalLabel,
        highlight,
        required,
        small,
        medium,
        large,
        xlarge,
        ...rest
      } = this.props as any;

      const inputContainer = { small, medium, large, xlarge };

      const errorId = `${this.id}-error`;

      return (
        <div
          className={cs(
            'form-group',
            styles['input'],
            className,
            highlight && styles.highlight,
            {
              'double-margin': doubleLabelMargin
            }
          )}
        >
          <Label
            id={`${this.id}-label`}
            htmlFor={this.id}
            required={required}
            strong={strong}
            doubleLabelMargin
            dataSelector={
              rest['data-selector'] && `${rest['data-selector']}-label`
            }
          >
            {label}
            {additionalLabel &&
              <span className={styles['label__additional']}>
                {additionalLabel}
              </span>}
          </Label>
          <div className={cs('input-wrapper', gridItemAdapter(inputContainer))}>
            <Comp
              id={this.id}
              name={name}
              invalid={invalid}
              aria-invalid={invalid}
              required={required}
              aria-describedby={errorId}
              {...rest}
            />
          </div>
          <div
            id={errorId}
            className="form-group"
            aria-hidden={!invalid}
            role="alert"
          >
            {invalid &&
              errorMessage &&
              <Error
                dataSelector={
                  (rest['data-selector'] && `${rest['data-selector']}-error`) ||
                  errorDataSelector
                }
                errorMessage={errorMessage}
              />}
          </div>
        </div>
      );
    }
  };
}
```

And the test looks like this:

```
const FormSelect = FormControl(Select);

const wrap = (props?) =>
  mount(<FormSelect options={[1, 2, 3]} label="label" {...props} />);

describe('FormControl', () => {
  it('should add aria tags for invalid', () => {
    const wrapper = wrap({ invalid: true });

    expect(wrapper.find('[aria-invalid=true]')).toHaveLength(1);
  });
```

**Comment by ljharb** [comment]
@dagda1 what does `wrapper.find('[aria-invalid=true]').debug()` say tho?

**Comment by dagda1** [comment]
sorry, it returns:

```
    console.log src/components/hoc/FormControl/FormControl.test.tsx:51
      <Component id="ctrl5" name={[undefined]} invalid={true} aria-invalid={true} required={[undefined]} aria-describedby="ctrl5-error" options={{...}} defaultLabel="" filterOptions={[Function]}>
        <div className={[undefined]}>
          <select value={[undefined]} className="undefined" id="ctrl5" name={[undefined]} aria-invalid={true} required={[undefined]} aria-describedby="ctrl5-error">
            <option value="" />
            <option>
              1
            </option>
            <option>
              2
            </option>
            <option>
              3
            </option>
          </select>
        </div>
      </Component>


      <select value={[undefined]} className="undefined" id="ctrl5" name={[undefined]} aria-invalid={true} required={[undefined]} aria-describedby="ctrl5-error">
        <option value="" />
        <option>
          1
        </option>
        <option>
          2
        </option>
        <option>
          3
        </option>
      </select>
```

**Comment by dagda1** [comment]
I have a number of tests that are returning 2 for each query and the common denominator of all of them is that I am querying by attribute.

For example in another test:

`console.log(wrapper.find('button').debug());`

returns 
```
      <button className="" type="button" onClick={[Function]} disabled={[undefined]} data-selector="cookie-banner-close">
        Close
      </button>
```

but `console.log(wrapper.find('[data-selector="cookie-banner-close"]'));`

returns

```

"<Component type="button" onClick={[Function]} data-selector="cookie-banner-close">
  <button className="" type="button" onClick={[Function]} disabled={[undefined]} data-selector="cookie-banner-close">
    Close
  </button>
</Component>


<button className="" type="button" onClick={[Function]} disabled={[undefined]} data-selector="cookie-banner-close">
  Close
</button>
```

**Comment by dagda1** [comment]
Is there an extra check on `nodeType` needed in `matchAttributeSelector` that checks for `nodeType === 'host'`? or `nodeType !== 'function'`

**Comment by ljharb** [comment]
No (you can inspect that in .findWhere tho); that’s what `.hostNodes()` is for. That debug output has one React component node and one host node; try `wrapper.find('[aria-invalid=true]').hostNodes().debug()`

**Comment by dagda1** [comment]
`wrapper.find('[aria-invalid=true]').hostNodes().debug()` ouputs:

```
     <select value={[undefined]} className="undefined" id="ctrl5" name={[undefined]} aria-invalid={true} required={[undefined]} aria-describedby="ctrl5-error">
        <option value="" />
        <option>
          1
        </option>
        <option>
          2
        </option>
        <option>
          3
        </option>
      </select>
```

This is what I am after but it feels weird that:

when I do find with a non-attribute selector, e.g. `wrapper.find("select").debug()`, I get:

```
<select value={[undefined]} className="undefined" id="ctrl5" name={[undefined]} aria-invalid={true} required={[undefined]} aria-describedby="ctrl5-error">
        <option value="" />
        <option>
          1
        </option>
        <option>
          2
        </option>
        <option>
          3
        </option>
      </select>
```

But when I call the same method (`find`) with an attribute type of query, I get extra results and have to call `hostNodes` which is not something I have ever heard of before to get the results I require.

Is this expected behaviour?

**Comment by dagda1** [comment]
I'd like to have a go at `fixing` this but I guess I first need to know if it is broke :).

**Comment by ljharb** [comment]
@dagda1 there's nothing to fix; it's expected behavior. `.find('select')` finds tags named "select", which of course are only host nodes. `.find('[foo]')` finds props named "foo", which might be on a host node, OR on a custom component.

In other words, it's *supposed* to search both component props and HTML element attributes (because they're all props).

**Comment by dagda1** [comment]
OK, thanks for the help but I've been using enzyme for a while and I've never encountered this until the upgrade and `hostNodes` is something and I could be wrong that not many will know about.

**Comment by ljharb** [comment]
Right, it's new in v3.

Would you like to submit a PR that adds it to the migration guide?

**Comment by dagda1** [comment]
Yes sir.  I'm on it.

**Comment by dagda1** [comment]
I have created #1408

**Comment by duro** [comment]
@ljharb I ran into this issue today, and luckily found this thread. I'm wondering if you wouldn't mind providing just a little more background on the use case that this change was made to support. As I sit presently, this seems like a mighty strange behavior with little justification for existing.

**Comment by ljharb** [comment]
@duro it makes the adapter pattern more generic, and smooths out API inconsistencies. In other words, in v2, some APIs implicitly filtered to host nodes, and some didn't. This change merely requires you to explicitly filter to host nodes if that's the thing you want, which makes your tests more explicit, and allows greater flexibility in adapters (which might not be using React at all).

**Comment by ljharb** [comment]
This should be resolved by #1408.
